File: script/resourceManager/ResourceManager.js
define(["jquery", "resourceManager/Model", "resourceManager/Shader", "resourceManager/Texture"],
    function ($, Model, Shader, Texture) {
        "use strict";

        /**
         *
         * @constructor
         * @param gl OpenGL context
         */
        var ResourceManager = function (gl) {
            this.shaders = {};
            this.models = {};
            this.textures = {};
            this.gl = gl;
        };

        /**
         * Possible resource types.
         * @enum {string}
         */
        ResourceManager.ResourceType = {
            SHADER: "shader",
            MODEL: "model",
            TEXTURE: "texture"
        };

        /**
         *
         * @param {{url:?string, id:?string}} resourceId
         * @param {ResourceManager.ResourceType} resourceType
         * @param {function(resource)} callback
         * @returns {*}
         */
        ResourceManager.prototype.getResource = function (resourceId, resourceType, callback) {
            if (typeof resourceId !== "object" || !resourceId.id && !resourceId.url) {
                throw "Resource ID must be object.";
            }

            if (resourceType === ResourceManager.ResourceType.MODEL) {
                if (!this.models[resourceId.id] && resourceId.url) {
                    this.loadModel(resourceId.url, callback);
                } else {
                    callback(this.models[resourceId.id]);
                }
            } else if (resourceType === ResourceManager.ResourceType.SHADER) {
                if (!this.shaders[resourceId.id] && resourceId.url) {
                    this.loadShader(resourceId.url, callback);
                } else {
                    callback(this.shaders[resourceId.id]);
                }
            } else if (resourceType === ResourceManager.ResourceType.TEXTURE) {
                if (!this.textures[resourceId.id] && resourceId.url) {
                    this.loadTexture(resourceId.url, callback);
                } else {
                    callback(this.textures[resourceId.id]);
                }
            }
        };

        ResourceManager.prototype.loadModel = function (url, callback) {
            var thi$ = this;
            $.getJSON(url,
                function (jModel) {
                    thi$.getResource(jModel.shader, ResourceManager.ResourceType.SHADER, function (oShader) {
                        thi$.models[jModel.id] = new Model(jModel, oShader, thi$.gl);
                        callback(thi$.models[jModel.id]);
                    });
                }
            );
        };

        ResourceManager.prototype.loadShader = function (url, callback) {
            var thi$ = this;
            var baseUrl = url.split("/");
            baseUrl.pop();
            baseUrl = baseUrl.join("/") + "/";

            $.getJSON(
                url,
                function (jShader) {
                    $.get(baseUrl + jShader.vertex,
                        function (sVertexSource) {
                            $.get(baseUrl + jShader.fragment,
                                function (sFragmentSource) {
                                    thi$.shaders[jShader.id] = new Shader(jShader, sVertexSource, sFragmentSource, thi$.gl);
                                    callback(thi$.shaders[jShader.id]);
                                }, "text");
                        },
                        "text");
                }
            );
        };

        ResourceManager.prototype.loadTexture = function (url) {

        };

        return ResourceManager;
    }
);

File: lib/resource_management/resource_manager.dart
part of resource_management;

class ResourceManager {
  static const int SHADER = 1;
  static const int MODEL = 2;

  gl.RenderingContext context;

  Map<String, Shader> shaders = new Map();
  Map<String, Model> models = new Map();

  ResourceManager(gl.RenderingContext context) : this.context = context;

  /**
   * Returns the resource specified by [resourceId] and/or [resourceURL]. If the resource is loaded it is returned by
   * ID, otherwise [resourceURL] is used to load it.
   */

  Resource getResource(int resourceType,
                       {String resourceId, String resourceURL}) {

    if (resourceType == SHADER) {

      if (resourceId != null && shaders.containsKey(resourceId)) {
        return shaders[resourceId];
      }

      return this.loadShader(resourceURL);

    } else if (resourceType == MODEL) {

      if (resourceId != null && models.containsKey(resourceId)) {
        return models[resourceId];
      }

      return this.loadModel(resourceURL);
    } else {
      throw new Exception('Invalid resource type.');
    }
  }

  Shader loadShader(String shaderURL) {
    HttpRequest shaderRequest = new HttpRequest();
    shaderRequest.open('GET', shaderURL, async:false);
    shaderRequest.send();

    String shaderJSONString = shaderRequest.responseText;
    Map shaderJSON = JSON.decode(shaderJSONString);

    var baseURL = shaderURL.split('/')
      ..removeLast();
    baseURL = baseURL.join('/');

    HttpRequest vertexRequest = new HttpRequest();
    vertexRequest.open('GET', baseURL + '/' + shaderJSON['vertex'],
    async:false);
    vertexRequest.send();
    HttpRequest fragmentRequest = new HttpRequest();
    fragmentRequest.open('GET', baseURL + '/' + shaderJSON['fragment'],
    async:false);
    fragmentRequest.send();

    Shader shader = new Shader(shaderJSON, vertexRequest.responseText,
    fragmentRequest.responseText, this.context);
    this.shaders[shader.id] = shader;

    return shader;
  }

  Model loadModel(String modelURL) {
    HttpRequest modelRequest = new HttpRequest();
    modelRequest.open('GET', modelURL, async:false);
    modelRequest.send();

    String modelJSONString = modelRequest.responseText;
    Map modelJSON = JSON.decode(modelJSONString);

    Model model = new Model(modelJSON, this.getResource(ResourceManager.SHADER,
    resourceId:modelJSON['shader']['id'],
    resourceURL:modelJSON['shader']['url']), this.context);
    this.models[model.id] = model;

    return model;
  }
}